libraries/azure_backend.rb

```diff
@@ -0,0 +1,713 @@
+require 'backend/azure_require'
+
+ENV_HASH = ENV.map { |k, v| [k.downcase, v] }.to_h
+
+# Base class for Azure resources.
+#
+# Provides:
+#  - Connection to Azure Rest API.
+#  - Parameter validation.
+#  - Short description of resources. This includes resource_id.
+#  - Detailed (long) description of resources upon providing a resource_id.
+#  - Latest or default api version for a resource provider endpoint.
+#  - Rescuing invalid api version error by using the suggested api version in the error message.
+#  - Creating resource methods dynamically.
+#
+class AzureResourceBase < Inspec.resource(1)
+  def initialize(opts = {})
+    raise ArgumentError, 'Parameters must be provided in an Hash object.' unless opts.is_a?(Hash)
+    @opts = opts
+
+    # Populate client_args to specify AzureConnection
+    #
+    # The valid client args (all of them are optional):
+    #  - endpoint: [String] azure_cloud (default), azure_china_cloud, azure_us_government, azure_german_cloud
+    #  - azure_retry_limit: [Integer] Maximum number of retries (default - 2)
+    #  - azure_retry_backoff: [Integer] Pause in seconds between retries (default - 0)
+    #  - azure_retry_backoff_factor: [Integer] The amount to multiply each successive retry's interval amount
+    #     by in order to provide back-off (default - 1)
+    @client_args = {}
+    # If not provided, the endpoint will be the Global Cloud portal.
+    # https://azure.microsoft.com/en-gb/global-infrastructure/
+    @client_args[:endpoint] = @opts[:endpoint] || ENV_HASH['endpoint'] || 'azure_cloud'
+    unless AzureEnvironments::ENDPOINTS.key?(@client_args[:endpoint])
+      raise ArgumentError, "Invalid endpoint: `#{@client_args[:endpoint]}`."\
+        " Expected one of the following options: #{AzureEnvironments::ENDPOINTS.keys}."
+    end
+    # Replace endpoint value with the content of AzureEnvironment instance.
+    # @type [Hash]
+    endpoint = AzureEnvironments.get_endpoint(@client_args[:endpoint])
+    @client_args[:endpoint] = endpoint
+    # Set HTTP client retry parameters, defining the timeout exception behavior, if provided.
+    @client_args[:azure_retry_limit] = @opts[:azure_retry_limit] || ENV_HASH['azure_retry_limit']
+    @client_args[:azure_retry_backoff] = @opts[:azure_retry_backoff] || ENV_HASH['azure_retry_backoff']
+    @client_args[:azure_retry_backoff_factor] = @opts[:azure_retry_backoff_factor] || ENV_HASH['azure_retry_backoff_factor']
```

Comment:
I assume these are set to nil when the opts and env_hash don't contain a value. Does the API permit a nil value?

Reply:
Ah nevermind I noticed this is being defaulted in the get_resources methods